### PR TITLE
feat: extend graphql with user roles

### PR DESCRIPTION
**Pull request by gmeben**
This PR extends the GraphQL endpoint with user roles.

### CODE REVIEW

File: backend/graphql/schema.graphql [hunk 1]
```diff
@@ -81,6 +91,7 @@ type User {
     username: String!
     created_at: DateTime!
     updated_at: DateTime
+    roles: [Role]
```
Comment (wreality) [inline backend/graphql/schema.graphql:94]:
```suggestion
    roles: [Role!]!
```

File: backend/tests/Feature/UserPermissionsTest.php [hunk 2]
```diff
@@ -126,4 +127,82 @@ public function testUserDoesNotHaveTestPermissionByAssignedRole()
...
+                'name' => 'Test User Role'
+            ]
+        ];
+        $getUserResponse->assertJsonPath("data.user.roles", $expected_array);
+    }
+
+    /**
+     * @return void
+     */
+    public function testRoleIsQueryableFromUserCreatedViaFactory()
```
Comment (wreality) [inline backend/tests/Feature/UserPermissionsTest.php:180]: And now I see that you have one created by the factory as well.  I guess I'm curious what we're testing differently btw these two tests.

**Review by wreality — CHANGES_REQUESTED** [review]

File: backend/tests/Feature/UserPermissionsTest.php [hunk 3]
```diff
@@ -126,4 +127,82 @@ public function testUserDoesNotHaveTestPermissionByAssignedRole()
...
+                    password: "KajSu8viptUrz&",
+                    username: "testusername",
+                    name: "Test Name"
+                }) {
+                    id
+                    name
+                    username
+                }
+            }'
+        );
```
[the remark this answers is not in the document]
Comment (gmeben) [reply backend/tests/Feature/UserPermissionsTest.php:149]: I think it's useful to test that the same information is available when creating users in different techniques.

File: backend/graphql/schema.graphql [hunk 4]
```diff
@@ -71,6 +73,14 @@ input CreateUserInput {
         @hash
 }
```
[the remark this answers is not in the document]
Comment (gmeben) [reply backend/graphql/schema.graphql:75]: Added

File: backend/graphql/schema.graphql [hunk 5]
```diff
@@ -14,6 +14,8 @@ type Query {
 
     "Validate details of a new user.  Returns true if supplied user fields would be valid."
     validateNewUser(user: ValidateNewUserInput): Boolean @returnTrue
+
```
[the remark this answers is not in the document]
Comment (gmeben) [reply backend/graphql/schema.graphql:17]: Added

File: backend/graphql/schema.graphql (hunk 1 above)
Comment (gmeben) [reply backend/graphql/schema.graphql:98]: Roles are optional at the moment. I suspect we'll have to do more work elsewhere to make them required before I can justify designating the return of null value here as an invalid request.

File: backend/tests/Feature/UserPermissionsTest.php (hunk 2 above)
Comment (gmeben) [reply backend/tests/Feature/UserPermissionsTest.php:180]: It's a sanity check. If a user was created via User factory or via Lighthouse, we want to ensure that the data exists and functions the same regardless.

File: backend/graphql/schema.graphql (hunk 1 above)
Comment (gmeben) [reply backend/graphql/schema.graphql:98]: I just tried querying a user that has no role with `[Role!]!` set in the schema and observed no different to `[Role]` so now I'm not even sure what designating this as a non-nullable field is actually doing.

Comment (wreality) [reply backend/graphql/schema.graphql:98]: We don't currently return null, we return an empty array. 

https://stackoverflow.com/questions/46770501/graphql-non-nullable-array-list/46771890

File: backend/tests/Feature/UserPermissionsTest.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (gmeben) [reply backend/tests/Feature/UserPermissionsTest.php]: This test was removed

File: backend/tests/Feature/UserPermissionsTest.php (hunk 2 above)
Comment (gmeben) [reply backend/tests/Feature/UserPermissionsTest.php]: The duplicate test was removed.

**Review by wreality — APPROVED** [review]

**Pull request merged**